File: communication_protocol/communication_module.py
import asyncio
from communication_protocol.message import connect_message, health_check_message
import machine  # type: ignore
import network  # type: ignore
import usocket  # type: ignore
from ucollections import deque  # type: ignore
from communication_protocol.communication_protocol import DeviceMessage
from communication_protocol.message_event import MessageEvent
from communication_protocol.message_type import MessageType


class CommunicationModule:
    def __init__(
        self, host_name: str, host_port: int, ssid: str, password: str, fun: str
    ):
        self.ssid = ssid
        self.password = password
        self.fun = fun
        self.to_server_queue = deque([], 1000)
        self.from_server_queue = deque([], 1000)
        self.wlan = network.WLAN(network.STA_IF)
        self.host_name = host_name
        self.host_port = host_port
        self.send_data_event = asyncio.Event()
        self.mac = ":".join(["{:02x}".format(b) for b in self.wlan.config("mac")])
        self.socket = None

    def is_connected(self) -> bool:
        return self.wlan.isconnected()

    def get_message(self) -> DeviceMessage | None:
        if self.from_server_queue:
            return self.from_server_queue.popleft()
        return None

    def extract_method_name_from_message(self, message: DeviceMessage) -> str:
        name = message.message_event.lower()
        type = message.message_type.lower()
        return f"_{name}_{type}"

    def send_message(self, message: DeviceMessage) -> None:
        self.to_server_queue.append(message)
        self.send_data_event.set()

    def get_mac(self) -> str:
        return self.mac

    async def start(self) -> None:
        receive_from_router_task = None
        send_to_router_task = None
        send_ping_task = None
        send_health_check_task = None
        while True:
            await self._connect_to_network()
            try:
                addr_info = usocket.getaddrinfo(self.host_name, self.host_port)[0][-1]
                if self.socket:
                    self.socket.close()
                if receive_from_router_task:
                    receive_from_router_task.cancel()
                if send_to_router_task:
                    send_to_router_task.cancel()
                if send_ping_task:
                    send_ping_task.cancel()
                if send_health_check_task:
                    send_health_check_task.cancel()

                self.socket = usocket.socket(usocket.AF_INET, usocket.SOCK_STREAM)
                self.socket.connect(addr_info)
                message: DeviceMessage = self.get_connect_message()
                self.socket.send(message.to_json().encode())
                receive_from_router_task = asyncio.create_task(
                    self._receive_from_router()
                )
                send_to_router_task = asyncio.create_task(self._send_to_router())
                send_ping_task = asyncio.create_task(self._send_ping())
                send_health_check_task = asyncio.create_task(self._send_health_check())

                await asyncio.gather(
                    receive_from_router_task,
                    send_to_router_task,
                    send_ping_task,
                    send_health_check_task,
                )
            except Exception as e:
                print(e)
                print("Resetuje połączenie")

    async def _receive_from_router(self) -> None:
        self.socket.setblocking(False)
        while True:
            try:
                data = self.socket.recv(1024)
                if data == b"P":
                    continue
                data = data.decode("utf-8")
                data: DeviceMessage = DeviceMessage.from_json(data)
                self.from_server_queue.append(data)
            except OSError as e:
                if e.args[0] not in (11, 35):
                    print(e)
                    raise
            await asyncio.sleep(0.1)

    async def _send_to_router(self) -> None:
        while True:
            if not self.to_server_queue:
                await self.send_data_event.wait()
                self.send_data_event.clear()
                continue
            message: DeviceMessage = self.to_server_queue.popleft()
            message.device_id = self.mac
            message = message.to_json()
            self.socket.sendall(message.encode())

    async def _connect_to_network(self):
        if not self.wlan.isconnected():
            self.wlan.active(True)
            self.wlan.connect(self.ssid, self.password)
            print("Trwa łączenie z siecią wifi...")
            count = 0
            while not self.wlan.isconnected():
                print(".")
                count += 1
                if count == 5:
                    machine.reset()
                await asyncio.sleep(1)
            print("Adres IP:", self.wlan.ifconfig()[0])

    async def _send_health_check(self) -> None:
        while True:
            self.to_server_queue.append(
                health_check_message(self.mac, self.wlan.status("rssi"))
            )
            self.send_data_event.set()
            await asyncio.sleep(120)

    async def _send_ping(self) -> None:
        while True:
            self.socket.sendall(b"P")
            await asyncio.sleep(1)

    def get_connect_message(self) -> DeviceMessage:
        return connect_message(self.mac, self.fun, self.wlan.status("rssi"))

File: arduino/lamp_esp8266_12E/src/CommunicationProtocol/communication_module.cpp
#include <Arduino.h>
#include "communication_module.h"
#include "settings.h"

CommunicationModule::CommunicationModule()
  : host_name(SERVER_IP), host_port(SERVER_PORT), ssid(SSID), password(PASSWORD), fun(DEVICE_FUNCTION){
  mac = WiFi.macAddress();

}

bool CommunicationModule::is_connected() {
  return WiFi.status() == WL_CONNECTED;
}

DeviceMessage* CommunicationModule::get_message() {
  if (from_server_count > 0) {
    DeviceMessage* msg = from_server_queue[0];
    for (int i = 0; i < from_server_count - 1; i++) {
      from_server_queue[i] = from_server_queue[i + 1];
    }
    from_server_count--;
    return msg;
  }
  return nullptr;
}

String CommunicationModule::extract_method_name_from_message(DeviceMessage& message) {
  return "_" + message.message_event + "_" + message.message_type;
}

void CommunicationModule::send_message(DeviceMessage message) {
  if (to_server_count < MAX_QUEUE_SIZE) {
    to_server_queue[to_server_count++] = new DeviceMessage(message);
  }
}

String CommunicationModule::get_mac() {
  return mac;
}

void CommunicationModule::start() {
  connect_to_network();
  if (!client.connected()) {
    if (client.connect(host_name.c_str(), host_port)) {
      DeviceMessage connect_msg = get_connect_message();
      client.println(connect_msg.toJson());
    }
  }
  send_to_router();
  receive_from_router();
  send_health_check();
}

void CommunicationModule::connect_to_network() {
  if (WiFi.status() != WL_CONNECTED) {
    WiFi.begin(ssid.c_str(), password.c_str());
    while (WiFi.status() != WL_CONNECTED) {
      delay(1000);
    }
  }
}

void CommunicationModule::receive_from_router() {
  if (client.connected() && client.available()) {
    client.setTimeout(100);
    String data = client.readStringUntil('\n');
    if (from_server_count < MAX_QUEUE_SIZE) {
      from_server_queue[from_server_count++] = new DeviceMessage(DeviceMessage::fromJson(data));
    }
  }
}

void CommunicationModule::send_to_router() {
  if (to_server_count > 0 && client.connected()) {
    DeviceMessage* msg = to_server_queue[0];
    for (int i = 0; i < to_server_count - 1; i++) {
      to_server_queue[i] = to_server_queue[i + 1];
    }
    to_server_count--;
    client.println(msg->toJson());
    delete msg;
  }
}

void CommunicationModule::send_health_check() {
  static unsigned long last_health_check = 0;
  if (millis() - last_health_check >= HEALT_CHECK_INTERVAL) {
    JsonDocument payload;
    payload["wifi_strength"] = WiFi.RSSI();
    send_message(DeviceMessage(String(millis()), "health_check", "request", mac, payload));
    last_health_check = millis();
  }
}

DeviceMessage CommunicationModule::get_connect_message() {
  JsonDocument payload;
  payload["fun"] = fun;
  payload["wifi_strength"] = WiFi.RSSI();
  return DeviceMessage(String(millis()), "device_connect", "request", mac, payload);
}